### PR TITLE
feat: Migration script for LDAP sync

### LINKED ISSUE #88: Migration executable for encoding old IDs
Write migration command for deployed instances of ldap-sync
- in Zitadel, currently the ids are either not encoded or base64 encoded, so we need to go through every instance and appropriately re-encode it in the new way
- this should be created as a secondary entry point with different binary within ldap-sync

More info [here](https://github.com/famedly/famedly-sync/pull/79#discussion_r1830209026).

### PR DESCRIPTION
**Pull request by jannden**
Closes #88

### CODE REVIEW

File: src/bin/migrate.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	tracing::info!("Starting migration");
+	tracing::debug!("Old external IDs will be base64 decoded and re-encoded as hex");
+	tracing::debug!("Note: External IDs are stored in the nick_name field of the user's profile in Zitadel, often referred to as uid.");
+
+	// Zitadel
+	let zitadel_config = config.zitadel.clone();
+	let mut zitadel = Zitadel::new(zitadel_config.url, zitadel_config.key_file).await?;
+
+	// Get all users
+	let mut stream = zitadel.list_users(
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:45]: Should we reuse the method from the ldap source for this? Then we don't have to do the user conversion here as well.

File: src/bin/migrate.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		ListUsersRequest::new(vec![
+			SearchQuery::new().with_type_query(TypeQuery::new(Userv2Type::Human))
+		])
+		.with_asc(true)
+		.with_sorting_column(UserFieldName::NickName),
+	)?;
+
+	// Process each user
+	while let Some(user) = stream.next().await {
+		let user_id = user.user_id().ok_or_else(|| anyhow!("User lacks ID"))?;
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:55]: Even if we don't reuse the method for getting the stream, we should probably reuse the user parsing as much as possible.

I wouldn't suggest doing so if we need to add options to the type to avoid parsing specific fields, of course, I'm not in the codebase enough right now to spot how awkward this would be with the existing types, and how much they could be redesigned to fit this purpose as well.

File: src/bin/migrate.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		let given_name = profile.given_name();
+		let family_name = profile.family_name();
+		let display_name = profile.display_name();
+
+		let Some(nick_name) = profile.nick_name() else {
+			tracing::warn!(user_id = %user_id, "User lacks nick_name (external uid), skipping");
+			continue;
+		};
+
+		tracing::info!(user_id = %user_id, old_uid = %nick_name, "Starting migration for user");
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:68]: This is fine, but the user struct deliberately implements display to prevent accidental future leaks of PII in logs. If we don't use the user struct, we should maybe leave a comment here to make sure nobody is tempted to start printing email addresses in the future.

File: src/bin/migrate.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+		let new_external_id = if nick_name.is_empty() {
+			// Keep empty string as is, will skip it later
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:71]: Why do we not just end early here? There are three cases, this would be cleaner as a `match`.

File: src/bin/migrate.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+			nick_name.to_string()
+		} else {
+			// First check if it's valid hex
+			if nick_name.chars().all(|c| c.is_ascii_hexdigit()) && nick_name.len() % 2 == 0 {
+				// If valid hex, keep as is
+				tracing::debug!(  user_id = %user_id, old_uid = %nick_name,"Valid hex uid detected, keeping as is");
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:77]: The valid set of characters for a hex-encoded string is a subset of the ones for a base64 encoded string. As such, while unlikely, this could end up mis-identifying some base64-encoded IDs as hex encoded.

We can still do this kind of check, but we should probably employ it as a heuristic (that can be overridden) on the whole database instead of per-record.

File: src/bin/migrate.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+				nick_name.to_string()
+			} else {
+				// Try base64, if fails use plain text
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:80]: Hmm, IDs which happen to be valid base64 will be "decoded" in this case. I realize we're working with spotty data here, because of how we encoded the IDs in the first place, but is there anything we can do to improve this?

Notably, due to the original implementation, the database should be either fully base64-encoded, or fully string-encoded. Maybe again a full-database heuristic to reduce the chances of failure here?

File: src/bin/migrate.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        new_uid = %new_external_id,
+        "Skipping user due to empty uid");
+			continue;
+		}
+
+		// Update uid (external ID, nick_name) in Zitadel
+		let mut request = UpdateHumanUserRequest::new();
+		request.set_profile(
+			SetHumanProfile::new(
+				given_name.ok_or_else(|| anyhow!("User lacks given name"))?.to_owned(),
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:106]: Are these mandatory? Can't we just update the ID?

File: tests/e2e.rs [hunk 8]
```diff
@@ -1396,6 +1397,112 @@ async fn test_e2e_sso_linking() {
 	);
 }
 
+#[test(tokio::test)]
+#[test_log(default_log_filter = "debug")]
+async fn test_e2e_migrate_base64_id() {
```
Comment (tlater-famedly) [inline tests/e2e.rs:1402]: Can these tests be put in a separate file?

File: tests/e2e.rs [hunk 9]
```diff
@@ -1396,6 +1397,112 @@ async fn test_e2e_sso_linking() {
 	);
 }
```
Comment (tlater-famedly) [inline tests/e2e.rs:1399]: We should add tests for the edge cases I mention further up; IDs that happen to be valid base64/hex, but aren't intended to be treated as such originally.

**Review by tlater-famedly — CHANGES_REQUESTED** [review]

File: src/bin/migrate.rs [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+				family_name.ok_or_else(|| anyhow!("User lacks family name"))?.to_owned(),
+			)
+			.with_display_name(
+				display_name.ok_or_else(|| anyhow!("User lacks display name"))?.to_owned(),
+			)
+			.with_nick_name(new_external_id.clone()),
+		);
+
+		zitadel.update_human_user(user_id, request).await?;
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:115]: We should respect the `DryRun` flag.

File: src/bin/migrate.rs (hunk 1 above)
Comment (jannden) [reply src/bin/migrate.rs:45]: Which method do you mean? `get_next_zitadel_user`?

Comment (tlater-famedly) [reply src/bin/migrate.rs:45]: This method: https://github.com/famedly/famedly-sync/blob/main/src/zitadel.rs#L91

File: src/bin/migrate.rs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+			}
+		};
+
+		// Skip empty uid
+		if new_external_id.is_empty() {
+			tracing::warn!(
+        user_id = %user_id,
+        old_uid = %nick_name,
+        new_uid = %new_external_id,
+        "Skipping user due to empty uid");
```
Comment (sirewix) [inline src/bin/migrate.rs:98]: broken indentation

File: src/zitadel.rs [hunk 12]
```diff
@@ -327,6 +310,24 @@ fn search_result_to_user(user: ZitadelUser) -> Result<User> {
...
+/// create collisions (although this is unlikely).
+///
+/// Only use this for Zitadel support.
+pub fn get_zitadel_encoded_id(external_id_bytes: Vec<u8>) -> Result<String> {
+	Ok(if let Ok(encoded_id) = String::from_utf8(external_id_bytes.clone()) {
+		encoded_id
+	} else {
+		BASE64_STANDARD.encode(external_id_bytes)
+	})
+}
```
Comment (sirewix) [inline src/zitadel.rs:329]:
```suggestion
pub fn get_zitadel_encoded_id(external_id_bytes: Vec<u8>) -> String {
	String::from_utf8(external_id_bytes.clone().unwrap_or_else(||
		BASE64_STANDARD.encode(external_id_bytes)
	)
}
```

File: src/bin/migrate.rs (hunk 1 above)
Comment (jannden) [reply src/bin/migrate.rs:45]: Took a bit of refactoring, but here we go.

File: src/zitadel.rs [hunk 13]
```diff
@@ -20,16 +20,20 @@ use zitadel_rust_client::{
...
+	get_next_zitadel_user,
 	user::User,
 	FeatureFlag,
 };
 
 /// The Zitadel project role to assign to users.
 const FAMEDLY_USER_ROLE: &str = "User";
 
+/// The number of users to sample for encoding detection
+const USER_SAMPLE_SIZE: usize = 10;
```
Comment (skomski) [inline src/zitadel.rs:32]: The sample size should be higher like 50+.

File: src/bin/migrate.rs [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+	// Use thresholds to determine encoding
+	let hex_ratio = f64::from(hex_count) / f64::from(total);
+	let base64_ratio = f64::from(base64_count) / f64::from(total);
+
+	if hex_ratio > 0.8 {
+		ExternalIdEncoding::Hex
+	} else if base64_ratio > 0.8 {
+		ExternalIdEncoding::Base64
+	} else {
+		ExternalIdEncoding::Plain
```
Comment (skomski) [inline src/bin/migrate.rs:102]: This function should be stricter. The migration tool should not continue if one format is not dominant and Plain should not be the default option when the others are not dominant. There should be like an `ambiguous` option.
Aside from that what happens if the migration fails in-between for whatever reason? Then it should not be possible to restart the migration tool because it could lead to a wrong outcome. First the database should be rolled-back, or?

File: src/bin/migrate.rs [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+	tracing::info!("Starting migration");
+	tracing::debug!("Old external IDs will be base64 decoded and re-encoded as hex");
+	tracing::debug!("Note: External IDs are stored in the nick_name field of the user's profile in Zitadel, often referred to as uid.");
+
+	// Zitadel
+	let mut zitadel = SyncZitadel::new(&config).await?;
+
+	// Detect external ID encoding based on a sample of users
+	let users_sample = zitadel.get_users_sample().await?;
+	let encoding = detect_database_encoding(users_sample);
```
Comment (skomski) [inline src/bin/migrate.rs:41]: Why does the tool continue when the dominant encoding is `hex`? Right now it would not convert the possible remaining base64 user_ids to hex because it would expect hex in `create_user_with_converted_external_id`, or?

**Review by skomski — CHANGES_REQUESTED** [review]
Added some questions that I encountered.

File: src/bin/migrate.rs (hunk 14 above)
Comment (jannden) [reply src/bin/migrate.rs:102]: Regarding the first point:
- Now an `ambiguous` option is introduced and `plain` is not default; when `ambiguous` is detected from the database heuristics, then it will go to check the particular case again regardless of the rest of the database.
- Fair point about the migration failing in-between, then a re-run would be tricky. After talking with @emgrav, for now it's acceptable to require any migration run to be done under controlled circumstances, i.e. with a db dump of zitadel taken right before, so we can just roll back; and in the long term, we should make it transactional.

File: src/bin/migrate.rs (hunk 15 above)
Comment (jannden) [reply src/bin/migrate.rs:41]: The tool continues so that we can get logs about the particular cases, ie. we will see how many users don't match the db-detected encoding. Yes, it would not convert the possible `base64` user IDs to `hex` if `hex` was detected in DB heuristics. That's by design. Any better solution? Keep in mind that the result of a `hex` encoding is a subset of characters of the `base64` encoding. That's what prompted us with @tlater-famedly to implement DB heuristics.

Comment (skomski) [reply src/bin/migrate.rs:41]: Yes makes sense. Just wondered why it would not stop because there is no need to migrate.

File: src/bin/migrate.rs (hunk 14 above)
Comment (skomski) [reply src/bin/migrate.rs]: Okay sounds good.

**Review by skomski — APPROVED** [review]
